README.md
# Base2<sup>15</sup>

binary-to-text encoding scheme that represents binary data in an unicode string format. Each unicode character represents 15 bits of binary data.

#### Example ####

| Encoding | Data | chararacters |
|:-:|:-:|:-:|
| Plain text | Base2e15 is awesome! | 20 |
| **Base2<sup>15</sup>** | **嗺둽嬖蟝巍媖疌켉溁닽壪** | **11** |
| Base64 | QmFzZTJlMTUgaXMgYXdlc29tZSE= | 27+1 |
 
## Mapping table

Every character represents 15 bits of data, except the last character, which is 7 or 15 bits.

| Binary | Unicode | Unicode Range Name |
|:-:|:-:|:-:|
| **15 bits mapping** | | |
| 0x0000 ~ 0x1935 | U+3480 ~ U+4DB5 | CJK Unified Ideographs Extension A |
| 0x1936 ~ 0x545B | U+4E00 ~ U+8925 | CJK Unified Ideographs |
| 0x545C ~ 0x7FFF | U+AC00 ~ U+D7A3 | Hangul Syllables |
| **7 bits mapping** | | |
| 0x00   ~ 0x7F | U+3400 ~ U+347F | CJK Unified Ideographs Extension A |

## Usage

A simple usage example in dart:
```dart
import 'dart:convert';
import 'package:base2e15/base2e15.dart';

main() {
  String msg = 'Base2e15 is awesome!';
  String encoded = Base2e15.encode(UTF8.encode(msg));
  String decoded = UTF8.decode(Base2e15.decode(encoded));
}
```

example in c:
```c
#include <stdio.h>
#include "base2e15.h"
int main(int argc, char *argv[])
{
    char c[] = "Base2e15 is awesome!";
    int len = strlen(c);

    int encodeLen = base2e15_encode_length(len) + 1; \\ one more char for the \0
    wchar_t* encoded = malloc(encodeLen * sizeof(wchar_t));
    base2e15_encode(c, len, encoded, encodeLen);
    
    // int buffLen = base2e15_decode_length(encoded, -1);
    int buffLen = 256; 
    char* decoded = malloc(buffLen * sizeof(char));
    int decodeLen = base2e15_decode(encoded, -1, decoded, buffLen);
    // if buffLen is not big enough, decodeLen will be -1
}
```

## Compare

| Compare | Base2<sup>15</sup> |  Base64 |
|:-:|:-:|:-:|
| bits per character | **15** | 6 |
| bits per char width | **7.5 (15/2)** | 6 (6/1) |
| bits per UTF8 byte | 5 (15/3) | **6 (6/1)** |
| bits per UTF16 byte | **7.5 (15/2)** | 3 (6/2) |

## Why not Base2<sup>16</sup>?

The unicode range `CJK Unified Ideographs Extension B` contains 42711 characters (U+20000 ~ U+2A6D6), together with the characters used by Base2<sup>15</sup>, there are more than 65536 usable characters to encode 16 bits in each character.

However, font support for `CJK Unified Ideographs Extension B` is missing in most mobile devices and using this code range will also reduce the bits capacity in UTF8 and UTF16 encoding, since those characters require one more byte in UTF8 and 2 more bytes in UTF16.


dart/CHANGELOG.md
# Changelog
## 1.0.0

- Null safety changes

## 0.3.0

- Dart 2 support

## 0.2.0

- Mapping table change

## 0.1.1

- Fix README.md

## 0.1.0

- Initial version


dart/test/all_test.dart
// Copyright (c) 2015, Rick Zhou. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library base2e15.test;

import 'package:base2e15/base2e15.dart';
import 'dart:convert';
//import 'dart:math';

bool testEqual(Object a, Object b, String testName) {
  if (a == b) {
    print('$testName Passed!');
    return true;
  } else {
    print('$testName Failed, "$a" != "$b"');
    return false;
  }
}

main() {
  String msg = 'Base2e15 is awesome!';
  String encoded = Base2e15.encode(utf8.encode(msg));
  testEqual(encoded, '嗺둽嬖蟝巍媖疌켉溁닽壪', 'Encoding Test');
  String decoded = utf8.decode(Base2e15.decode(encoded));
  testEqual(decoded, msg, 'Decoding Test');
  String encoded2 = '~嗺둽嬖蟝巍媖疌123켉溁닽壪';
  String decoded2 = utf8.decode(Base2e15.decode(encoded2));
  testEqual(decoded2, msg, 'Malformed Decoding Test');

//  Random rng = new Random();
//  List bytes = new List(100);
//  for (int i = 0; i < 10000; ++i) {
//    int k = 50 + rng.nextInt(50);
//    for (int j = 0; j < k; ++j) {
//      bytes[j] = rng.nextInt(256);
//    }
//    String encoded = Base2e15.encode(bytes.sublist(0,k));
//    List newbytes = Base2e15.decode(encoded);
//    for (int j = 0; j < k; ++j) {
//      if (bytes[j] != newbytes[j]) {
//        print('Failed:$bytes');
//        print('      :$newbytes');
//        break;
//      }
//    }
//  }
}


c/base2e15.c
// Copyright (c) 2015, Rick Zhou. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

#include "base2e15.h"

/**
 * @param data is input data
 * @param len is data length in bytes
 * @param out is output memory
 * @param outlen is output size
 * @return output lenth used, -1 if output length is not enough
 */
int base2e15_encode(const void* data, int len, wchar_t* out, int outlen) {
    const unsigned char* input = (const unsigned char*)data;
    int outlenNeeded = (len * 8 + 14) / 15;
    if (outlen < outlenNeeded) {
        return -1;
    }
    if (outlen > outlenNeeded){
        out[outlenNeeded] = 0;
    }
    int bn = 15; // bit needed
    wchar_t bv = 0; // bit value

    const unsigned char* pEnd = input + len;
    wchar_t *pOut = out;
    const unsigned char* p;
    for (p = input; p < pEnd; ++p) {
      if (bn > 8) {
        bv = (bv << 8) | *p;
        bn -= 8;
      } else {
        bv = ((bv << bn) | (*p >> (8 - bn))) & 0x7FFF;
        if (bv < 0x1936) {
          *(pOut++) = bv + 0x3480;
        } else if (bv < 0x545C) {
          *(pOut++) = bv + 0x34CA;
        } else {
          *(pOut++) = bv + 0x57A4;
        }
        bv = *p;
        bn += 7;
      }
    }
    if (bn != 15) {
      if (bn > 7) {
        *(pOut++) = ((bv << (bn - 8)) & 0xFF) + 0x3400;
      } else {
        bv = (bv << bn) & 0x7FFF;
        if (bv < 0x1936) {
          *(pOut++) = bv + 0x3480;
        } else if (bv < 0x545C) {
          *(pOut++) = bv + 0x34CA;
        } else {
          *(pOut++) = bv + 0x57A4;
        }
      }
    }
    return outlenNeeded;
}

/**
 * @param length is the length in bytes for the data
 * @return length of the base2e15 needed, not including the trailing \0
 */
int base2e15_encode_length(int len) {
    return (len * 8 + 14) / 15;
}


/**
 * @param data is input data
 * @param len is data length, -1: auto detect
 * @return output lenth needed
 */
int base2e15_decode_length(const wchar_t* data, int len) {
    int count = 0;
    int bn = 8; // bit needed
    const wchar_t* dataEnd = data + len;
    const wchar_t* code;
    for (code = data; len == -1 || code < dataEnd; ++code) {
      if (*code == 0) {
        break;
      }
      if (*code > 0x33FF && *code < 0xD7A4) {
        if (*code > 0xABFF) {
        } else if (*code > 0x8925) {
          continue; // invalid range
        } else if (*code > 0x4DFF) {
        } else if (*code > 0x4DB5) {
          continue; // invalid range
        } else if (*code > 0x347F) {
        } else {
          ++count;
          break; // last 7 bit data received, break
        }
        ++count;

        bn -= 7;
        if (bn < 1) {
          ++count;
          bn += 8;
        }
      }
    }
    return count;
}

/**
 * @param data is input data
 * @param len is data length, -1: auto detect
 * @param out is output memory
 * @param outlen is output size in bytes
 * @return output lenth used, -1 if output length is not enough
 */
int base2e15_decode(const wchar_t* data, int len, void* out, int outlen) {
    unsigned char* output = (unsigned char*)out;
    unsigned char* outputEnd = output + outlen;

    int bn = 8; // bit needed
    wchar_t bv = 0; // bit value
    wchar_t cv;
    const wchar_t* dataEnd = data + len;
    const wchar_t* code;
    for (code = data; len == -1 || code < dataEnd; ++code) {
      if (*code == 0) {
        break;
      }
      if (*code > 0x33FF && *code < 0xD7A4) {
        if (*code > 0xABFF) {
          cv = *code - 0x57A4;
        } else if (*code > 0x8925) {
          continue; // invalid range
        } else if (*code > 0x4DFF) {
          cv = *code - 0x34CA;
        } else if (*code > 0x4DB5) {
          continue; // invalid range
        } else if (*code > 0x347F) {
          cv = *code - 0x3480;
        } else {
          cv = *code - 0x3400;
          *(output++) = (bv << bn) | (cv >> (7 - bn));
          break; // last 7 bit data received, break
        }
        if (output >= outputEnd) {
            return -1;
        }
        *(output++) = (bv << bn) | (cv >> (15 - bn));
        bv = cv;
        bn -= 7;
        if (bn < 1) {
          if (output >= outputEnd) {
            return -1;
          }
          *(output++) = bv >> -bn;
          bn += 8;
        }
      }
    }
    return output - (unsigned char*)out;
}



c/base2e15.h
// Copyright (c) 2015, Rick Zhou. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

#ifndef __BASE2E15_H__
#define __BASE2E15_H__
#include <wchar.h>

/**
 * @param data is input data
 * @param len is data length in bytes
 * @param out is output memory
 * @param outlen is output size
 * @return output lenth used, -1 if output length is not enough
 */
int base2e15_encode(const void* data, int len, wchar_t* out, int outlen);

/**
 * @param length is the length in bytes for the data
 * @return length of the base2e15 needed, not including the trailing \0
 */
int base2e15_encode_length(int len);


/**
 * @param data is input data
 * @param len is data length, -1: auto detect
 * @param out is output memory
 * @param outlen is output size in bytes
 * @return output lenth used, -1 if output length is not enough
 */
int base2e15_decode(const wchar_t* data, int len, void* out, int outlen);

/**
 * @param data is input data
 * @param len is data length, -1: auto detect
 * @return output lenth needed
 */
int base2e15_decode_length(const wchar_t* data, int len);

#endif


java/src/com/base2e15/Base2e15.java
// Copyright (c) 2015, Rick Zhou. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.
package com.base2e15;

import java.util.Arrays;

class Base2e15 {
  public static String encode(byte[] byts) {
    int bn = 15; // bit needed
    int bv = 0; // bit value
    int outLen = (byts.length * 8 + 14) / 15;
    int[] out = new int[outLen];
    int pos = 0;
    for (int i = 0; i < byts.length; ++i) {
        byte byt = byts[i];
      if (bn > 8) {
        bv = (bv << 8) | byt;
        bn -= 8;
      } else {
        bv = ((bv << bn) | (byt >> (8 - bn))) & 0x7FFF;
        if (bv < 0x1936) {
          out[pos++] = bv + 0x3480;
        } else if (bv < 0x545C) {
          out[pos++] = bv + 0x34CA;
        } else {
          out[pos++] = bv + 0x57A4;
        }
        bv = byt;
        bn += 7;
      }
    }
    if (bn != 15) {
      if (bn > 7) { // need 8 bits or more, so has 7 bits or less
        out[pos++] = ((bv << (bn - 8)) & 0x7F) + 0x3400;
      } else {
        bv = (bv << bn) & 0x7FFF;
        if (bv < 0x1936) {
          out[pos++] = bv + 0x3480;
        } else if (bv < 0x545C) {
          out[pos++] = bv + 0x34CA;
        } else {
          out[pos++] = bv + 0x57A4;
        }
      }
    }
    return new String(out, 0, outLen);
  }

  public static byte[] decode(String input) {
    int bn = 8; // bit needed
    int bv = 0; // bit value
    int inputLen = input.length();
    int maxLen = (inputLen * 15 + 7) / 8;
    byte[] out = new byte[maxLen];
    int pos = 0;
    int cv;
    for (int i = 0; i < inputLen; ++i) {
      char code = input.charAt(i);
      if (code > 0x33FF && code < 0xD7A4) {
        if (code > 0xABFF) {
          cv = code - 0x57A4;
        } else if (code > 0x8925) {
          continue; // invalid range
        } else if (code > 0x4DFF) {
          cv = code - 0x34CA;
        } else if (code > 0x4DB5) {
          continue; // invalid range
        } else if (code > 0x347F) {
          cv = code - 0x3480;
        } else {
          cv = code - 0x3400;
          out[pos++] = (byte)((bv << bn) | (cv >> (7 - bn)));
          break; // last 8 bit data received, break
        }
        out[pos++] =(byte)((bv << bn) | (cv >> (15 - bn)));
        bv = cv;
        bn -= 7;
        if (bn < 1) {
          out[pos++] = (byte)(bv >> -bn);
          bn += 8;
        }
      }
    }
    return Arrays.copyOfRange(out, 0, pos);
  }
}

dart/lib/base2e15.dart
// Copyright (c) 2015, Rick Zhou. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

/// The base2e15 library.
/// Map 15 bits to unicode
/// 0x0000 ~ 0x1935 -> U+3480 ~ U+4DB5   CJK Unified Ideographs Extension A
/// 0x1936 ~ 0x545B -> U+4E00 ~ U+8925   CJK Unified Ideographs
/// 0x545C ~ 0x7FFF -> U+AC00 ~ U+D7A3   Hangul Syllables
/// 7 bits special case, only used by last character
///  0x00  ~  0x7F  -> U+3400 ~ U+347F   CJK Unified Ideographs Extension A
library base2e15;

import 'dart:typed_data';

class Base2e15 {
  static String encode(List<int> bytes,
      [int lineSize = 0, String? linePadding]) {
    List<int?> charCodes = encodeToCharCode(bytes);
    if (lineSize <= 0) {
      return new String.fromCharCodes(charCodes as Iterable<int>);
    }
    List rslt = [];
    int len = charCodes.length;
    for (int i = 0; i < len; i += lineSize) {
      int j = i + lineSize;
      if (j < len) {
        j = len;
      }
      if (linePadding != null) {
        rslt.add(
            '$linePadding${new String.fromCharCodes(charCodes.sublist(i, j) as Iterable<int>)}');
      } else {
        rslt.add(
            new String.fromCharCodes(charCodes.sublist(i, j) as Iterable<int>));
      }
    }
    return rslt.join('\n');
  }

  static List<int?> encodeToCharCode(List<int> bytes) {
    int bn = 15; // bit needed
    int bv = 0; // bit value
    int outLen = (bytes.length * 8 + 14) ~/ 15;
    List<int> out = new List<int>.filled(outLen, -1, growable: false);
    int pos = 0;
    for (int byte in bytes) {
      if (bn > 8) {
        bv = (bv << 8) | byte;
        bn -= 8;
      } else {
        bv = ((bv << bn) | (byte >> (8 - bn))) & 0x7FFF;
        if (bv < 0x1936) {
          out[pos++] = bv + 0x3480;
        } else if (bv < 0x545C) {
          out[pos++] = bv + 0x34CA;
        } else {
          out[pos++] = bv + 0x57A4;
        }
        bv = byte;
        bn += 7;
      }
    }
    if (bn != 15) {
      if (bn > 7) {
        // need 8 bits or more, so has 7 bits or less
        out[pos++] = ((bv << (bn - 8)) & 0x7F) + 0x3400;
      } else {
        bv = (bv << bn) & 0x7FFF;
        if (bv < 0x1936) {
          out[pos++] = bv + 0x3480;
        } else if (bv < 0x545C) {
          out[pos++] = bv + 0x34CA;
        } else {
          out[pos++] = bv + 0x57A4;
        }
      }
    }
    return out;
  }

  static Uint8List decode(String input) {
    int bn = 8; // bit needed
    int bv = 0; // bit value
    int maxLen = (input.length * 15 + 7) ~/ 8;
    Uint8List out = new Uint8List(maxLen);
    int pos = 0;
    int cv;
    for (int code in input.codeUnits) {
      if (code > 0x33FF && code < 0xD7A4) {
        if (code > 0xABFF) {
          cv = code - 0x57A4;
        } else if (code > 0x8925) {
          continue; // invalid range
        } else if (code > 0x4DFF) {
          cv = code - 0x34CA;
        } else if (code > 0x4DB5) {
          continue; // invalid range
        } else if (code > 0x347F) {
          cv = code - 0x3480;
        } else {
          cv = code - 0x3400;
          out[pos++] = (bv << bn) | (cv >> (7 - bn));
          break; // last 8 bit data received, break
        }
        out[pos++] = (bv << bn) | (cv >> (15 - bn));
        bv = cv;
        bn -= 7;
        if (bn < 1) {
          out[pos++] = bv >> -bn;
          bn += 8;
        }
      }
    }
    return out.sublist(0, pos);
  }
}
